src/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs

```diff
@@ -154,6 +154,84 @@ private ReadOnlyMemory<T> GetFirstBufferSlow(object startObject, bool isMultiSeg
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ReadOnlySpan<T> GetFirstSpan()
+        {
+            object startObject = _startObject;
+
+            if (startObject == null)
+                return default;
+
+            int startIndex = _startInteger;
+            int endIndex = _endInteger;
+
+            bool isMultiSegment = startObject != _endObject;
+
+            // The highest bit of startIndex and endIndex are used to infer the sequence type
+            // The code below is structured this way for performance reasons and is equivalent to the following:
+            // SequenceType type = GetSequenceType();
+            // if (type == SequenceType.MultiSegment) { ... }
+            // else if (type == SequenceType.Array) { ... }
+            // else if (type == SequenceType.String){ ... }
+            // else if (type == SequenceType.MemoryManager) { ... }
+
+            // Highest bit of startIndex: A = startIndex >> 31
+            // Highest bit of endIndex: B = endIndex >> 31
+
+            // A == 0 && B == 0 means SequenceType.MultiSegment
+            // Equivalent to startIndex >= 0 && endIndex >= 0
+            if ((startIndex | endIndex) >= 0)
+            {
+                ReadOnlySpan<T> span = ((ReadOnlySequenceSegment<T>)startObject).Memory.Span;
+                if (isMultiSegment)
+                {
+                    return span.Slice(startIndex);
+                }
+                return span.Slice(startIndex, endIndex - startIndex);
+            }
+            else
+            {
+                return GetFirstSpanFast(startObject, isMultiSegment);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private ReadOnlySpan<T> GetFirstSpanFast(object startObject, bool isMultiSegment)
```

Comment:
The "fast"/common path should not be in a non-inlined helper. It's the less common path that we should extract to a non-inlined method.

`SequenceType.Array` is on the fast path. The `SequenceType.MemoryManager` is on the slow path.

Reply:
Alternatively, inline the whole thing with the 4 if checks (combinations of `startIndex < 0`, `>= 0` and `endIndex <0`, `>= 0`).

Reply:
@ahsonkhan 
The current if else condition order is as follows (as also visible from the comments)

```
            // SequenceType type = GetSequenceType();
            // if (type == SequenceType.MultiSegment) { ... }
            // else if (type == SequenceType.Array) { ... }
            // else if (type == SequenceType.String){ ... }
            // else if (type == SequenceType.MemoryManager) { ... }
```
If I inline the code block of `GetFirstSpanFast` as-is into its calling function (into `GetFirstSpan`), the the order above still remains.
The slow path (MemoryManager is already at the bottom).

Are you asking for the order to change or is it sufficient for you if I just remove GetFirstSpanFast and inline the entire thing into GetFirstSpan?

Reply:
Alternative order is 
```
            // SequenceType type = GetSequenceType();
            // if (type == SequenceType.Array) { ... }
            // else if (type == SequenceType.String){ ... }
            // else if (type == SequenceType.MemoryManager) { ... }
            // else // (type == SequenceType.MultiSegment) { ... }
```

Reply:
Actually, you are likely correct.
Looking at previous changes in this space, it might be better to keep what you initially had - following the existing `GetFirstBuffer`/`GetFirstBufferSlow`:
https://github.com/dotnet/corefx/pull/33000

We should consider measuring the difference between the two approaches for the 4 sequence types to help decide what approach to take:
https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-corefx.md#benchmarking-new-api

I had shared some perf numbers previously but they are likely out-dated: https://github.com/dotnet/corefx/issues/33029#issuecomment-432942509

Let me know if you need any help running the perf tests.